exporter/sapmexporter/exporter.go

```diff
@@ -25,53 +25,115 @@ import (
 	"go.opentelemetry.io/collector/exporter/exporterhelper"
 	"go.opentelemetry.io/collector/translator/trace/jaeger"
 	"go.uber.org/zap"
+
+	"github.com/open-telemetry/opentelemetry-collector-contrib/internal/common/splunk"
 )
 
 // sapmExporter is a wrapper struct of SAPM exporter
 type sapmExporter struct {
 	client *sapmclient.Client
 	logger *zap.Logger
+	config *Config
 }
 
 func (se *sapmExporter) Shutdown(context.Context) error {
 	se.client.Stop()
 	return nil
 }
 
-func newSAPMTraceExporter(cfg *Config, params component.ExporterCreateParams) (component.TraceExporter, error) {
+func newSAPMExporter(cfg *Config, params component.ExporterCreateParams) (sapmExporter, error) {
 	err := cfg.validate()
 	if err != nil {
-		return nil, err
+		return sapmExporter{}, err
 	}
 
 	client, err := sapmclient.New(cfg.clientOptions()...)
 	if err != nil {
-		return nil, err
+		return sapmExporter{}, err
 	}
-	se := sapmExporter{
+	return sapmExporter{
 		client: client,
 		logger: params.Logger,
+		config: cfg,
+	}, err
+}
+
+func newSAPMTraceExporter(cfg *Config, params component.ExporterCreateParams) (component.TraceExporter, error) {
+	se, err := newSAPMExporter(cfg, params)
+	if err != nil {
+		return nil, err
 	}
+
 	return exporterhelper.NewTraceExporter(
 		cfg,
 		se.pushTraceData,
 		exporterhelper.WithShutdown(se.Shutdown))
 }
 
-// pushTraceData exports traces in SAPM proto and returns number of dropped spans and error if export failed
-func (se *sapmExporter) pushTraceData(ctx context.Context, td pdata.Traces) (droppedSpansCount int, err error) {
-	batches, err := jaeger.InternalTracesToJaegerProto(td)
-	if err != nil {
-		return td.SpanCount(), consumererror.Permanent(err)
+// tracesByAccessToken takes a pdata.Traces struct and will iterate through its ResourceSpans' attributes,
+// regrouping by any SFx access token label value if Config.AccessTokenPassthrough is enabled.  It will drop any
+// set token label in any case to prevent serialization.
+// It returns a map of newly constructed pdata.Traces keyed by access token, defaulting to empty string.
+func (se *sapmExporter) tracesByAccessToken(td pdata.Traces) map[string]pdata.Traces {
+	tracesByToken := make(map[string]pdata.Traces, 1)
+	resourceSpans := td.ResourceSpans()
+	for i := 0; i < resourceSpans.Len(); i++ {
+		resourceSpan := resourceSpans.At(i)
+		if resourceSpan.IsNil() {
+			// Invalid trace so nothing to export
```

Comment:
Not sure about hits. @pjanotti do we need to export empty Span slice to make sure our metrics are correct?

Reply:
This should be fine: we want the exported metrics to reflect the actual number of exported spans and since a nil span is not exported that is correct.

@rmfitzpatrick just double checking: this is just defensive programming, right? or did you encounter it in practice?

Reply:
Correct, I've not run into this being the case but wanted to ensure I wouldn't cause panics: https://github.com/open-telemetry/opentelemetry-collector/blob/a8db6271f90ad6877e77967cfaf288cda6e810bd/consumer/pdata/generated_trace.go#L172

Reply:
👍